### PR TITLE
feat: compute GCP projects parallelly

### PR DESCRIPTION
**Pull request by peb-peb**
## Description

- Parallelized GCP projects computation using the `multiprocessing.Pool()`.
- Improves tool runtime from ~120s to ~40s [66 % faster]. (for 3 projects running simultaneously each with a VM running on them)

## Changes Made

- `scanner.py` - shifted the main loop from `src/gcp_scanner/scanner.py` to here.
- `src/gcp_scanner/scanner.py` - make resource computation functions.
- `src/gcp_scanner/models.py` - make a `ProjectInfo` model to pass through each project iteration with required information for resource scanning.
- `src/gcp_scanner/arguments.py` - added the `worker_count` param to the CLI option.

## Checklist

- [x] implement `impersonation` over discovered service accounts.
- [ ] documentation for the above changes.
- [x] update relevant tests.

## Related Issues

#242 #245 

## Additional Notes

To run the above changes - `python3 scanner.py -o ../output -k ../output/sakeys -l INFO -wc 8`

**Screenshot for above implementation**-
image

### CODE REVIEW

**Comment by mshudrak** [comment]
This is great. Thanks for the PR. I will review it closer to weekend. Could you please fix tests in meanwhile? It seems like it fails acceptance test.

**Comment by peb-peb** [comment]
@mshudrak I've transferred the main loop to `scanner.main()`. The tests should be passing, but the error `FAILED src/gcp_scanner/test_acceptance.py::test_acceptance - AssertionError: assert 30 == 31` comes.

File: src/gcp_scanner/scanner.py [hunk 1]
```diff
@@ -129,166 +130,104 @@ def save_results(res_data: Dict, res_path: str, is_light: bool):
...
-    sa_results['service_account_chain'] = chain_so_far
-    sa_results['current_service_account'] = sa_name
-    # Add token scopes in the result
-    sa_results['token_scopes'] = credentials.scopes
-
-    project_list = CrawlerFactory.create_crawler(
-      'project_list',
-    ).crawl(
-      ClientFactory.get_client('cloudresourcemanager').get_service(credentials),
+def get_resources(project):
```
Comment (mshudrak) [inline src/gcp_scanner/scanner.py:133]: Please follow our function definition guideline. We usually have parameter types and function description

File: scanner.py [hunk 2]
```diff
@@ -14,5 +14,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
```
Comment (mshudrak) [inline scanner.py:17]: no need for extra line

File: src/gcp_scanner/arguments.py [hunk 3]
```diff
@@ -124,6 +124,12 @@ def arg_parser():
       dest='log_file',
       help='Save logs to the path specified rather than displaying in\
  console')
+  parser.add_argument(
+      '-wc',
+      '--worker-count',
+      default=1,
+      dest='worker_count',
+      help='Give Worker Count limit for parallelization of crawlers.')
```
Comment (mshudrak) [inline src/gcp_scanner/arguments.py:132]: I'd rephrase it to be just `Set limit for workers run in parallel`

File: src/gcp_scanner/models.py [hunk 4]
```diff
@@ -37,3 +37,31 @@ def __init__(self, sa_tuples: List[Tuple[str, Credentials, List[str]]]):
     self.service_account_queue = queue.Queue()
     for sa_tuple in sa_tuples:
       self.service_account_queue.put(sa_tuple)
+
+class ProjectInfo:
+  """A simple class to crawl resources with the required information for crawl
```
Comment (mshudrak) [inline src/gcp_scanner/models.py:42]: I'd rephrase to `A simple class to store project scan configuration`

File: src/gcp_scanner/scanner.py [hunk 5]
```diff
@@ -129,166 +130,104 @@ def save_results(res_data: Dict, res_path: str, is_light: bool):
...
-            logging.error('Failed to get token for %s',
-                          candidate_service_account)
-            logging.error(sys.exc_info()[1])
-
-      logging.info('Saving results for %s into the file', project_id)
-
-      save_results(sa_results, output_path, light_scan)
-      # Clean memory to avoid leak for large amount projects.
-      sa_results.clear()
+  # Iterate over discovered service accounts by attempting impersonation
```
Comment (mshudrak) [inline src/gcp_scanner/scanner.py:189]: I'd prefer that functionality to be implemented in this commit instead of just commenting it out. What we can basically do is to create a separate function that does impersonation and then call it from the loop starting at line 393. This way your build a list of available SAs and associated projects and scan them.

**Review by mshudrak — CHANGES_REQUESTED** [review]
Thanks for the PR. Looks pretty good so far. We need to finish impersonation component and then we can merge it.

File: src/gcp_scanner/scanner.py [hunk 6]
```diff
@@ -363,21 +302,7 @@ def infinite_defaultdict():
...
-  force_projects_list = list()
-  if args.force_projects:
-    force_projects_list = args.force_projects.split(',')
-
-  logging.basicConfig(level=getattr(logging, args.log_level.upper(), None),
-                      format='%(asctime)s - %(levelname)s - %(message)s',
-                      datefmt='%Y-%m-%d %H:%M:%S',
-                      filename=args.log_file, filemode='a')
-
+def get_sa_tuples(args):
```
Comment (mshudrak) [inline src/gcp_scanner/scanner.py:305]: A short description is needed too.

File: src/gcp_scanner/scanner.py (hunk 5 above)
Comment (peb-peb) [reply src/gcp_scanner/scanner.py:196]: So, we basically run two loops, first for the SA impersonation and the other for crawling.

Comment (peb-peb) [reply src/gcp_scanner/scanner.py:196]: Done

**Comment by peb-peb** [comment]
@mshudrak Ready for Review. 

I have implemented the `impersonation` and it works as expected. But, the tests are still failing.

**Comment by mshudrak** [comment]
I also noticed this error when I try to work with GCP SA key
```
023-08-15 21:37:53 - INFO - >> current service account: TBD
2023-08-15 21:37:53 - INFO - Retrieving projects list
Traceback (most recent call last):
  File "/usr/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_cffi_backend.FFI' object
Traceback (most recent call last):
  File "/usr/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_cffi_backend.FFI' object
```

File: src/gcp_scanner/scanner.py [hunk 7]
```diff
@@ -434,11 +389,116 @@ def main():
...
+        args.light_scan,
+        args.target_project,
+        scan_time_suffix,
+        sa_name,
+        credentials,
+        chain_so_far
+      )
+      project_queue.put(project_obj)
+      impersonate_service_accounts(
+        context,
```
Comment (mshudrak) [inline src/gcp_scanner/scanner.py:486]: How do you know the context is not copied here? Are we working with the same context above? We need to make sure that it is actually updated.

**Review by mshudrak — COMMENTED** [review]
I have a question. PTAL.

**Review by mshudrak — APPROVED** [review]
